homeassistant/components/trafikverket_train/sensor.py

```diff
@@ -0,0 +1,185 @@
+"""Train information for departures and delays, provided by Trafikverket."""
+
+from datetime import date, datetime, timedelta
+import logging
+
+from pytrafikverket import TrafikverketTrain
+import voluptuous as vol
+
+from homeassistant.components.sensor import PLATFORM_SCHEMA
+from homeassistant.const import CONF_API_KEY, CONF_NAME, CONF_WEEKDAY, WEEKDAYS
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity import Entity
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_TRAINS = "trains"
+CONF_FROM = "from"
+CONF_TO = "to"
+CONF_TIME = "time"
+
+ATTR_CANCELED = "canceled"
+ATTR_DELAY_TIME = "number_of_minutes_delayed"
+ATTR_PLANNED_TIME = "planned_time"
+ATTR_ESTIMATED_TIME = "estimated_time"
+ATTR_ACTUAL_TIME = "actual_time"
+ATTR_OTHER_INFORMATION = "other_information"
+ATTR_DEVIATIONS = "deviations"
+
+ICON = "mdi:train"
+SCAN_INTERVAL = timedelta(minutes=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Required(CONF_TRAINS): [{
+        vol.Required(CONF_NAME): cv.string,
+        vol.Required(CONF_TO): cv.string,
+        vol.Required(CONF_FROM): cv.string,
+        vol.Optional(CONF_TIME): cv.time,
+        vol.Optional(CONF_WEEKDAY, default=WEEKDAYS):
+            vol.All(cv.ensure_list, [vol.In(WEEKDAYS)])}]
+})
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Set up the departure sensor."""
+    httpsession = async_get_clientsession(hass)
+    train_api = TrafikverketTrain(httpsession, config.get(CONF_API_KEY))
+    sensors = []
+    station_cache = {}
+    for train in config.get(CONF_TRAINS):
+        try:
+            if train.get(CONF_FROM) in station_cache:
+                from_station = station_cache.get(train.get(CONF_FROM))
+            else:
+                from_station = await train_api.async_get_train_station(
+                    train.get(CONF_FROM))
+                station_cache[train.get(CONF_FROM)] = from_station
+
+            if train.get(CONF_TO) in station_cache:
```

Comment:
Looks like this block is a repetition of the block above besides replacing `CONF_FROM` with `CONF_TO`.

I suggest making another for loop that iterates over those keys.

Reply:
Having only CONF_TO and CONF_FROM would not be a problem to make a for loop instead, but we also assign the output to `from_station` and `to_station` variables and i tried with a dictionary.

This works also, but it's not very easy to read? ;)

```
    for train in config[CONF_TRAINS]:
        try:
            trainstops = {"from_station":[train.get(CONF_FROM), None],
                          "to_station":[train.get(CONF_TO), None]}

            for station in trainstops:
                _LOGGER.error(trainstops[station][0])
                if trainstops[station][0] in station_cache:
                    trainstops[station][1] = station_cache.get(trainstops[station][0])
                else:
                    trainstops[station][1] = await train_api.async_get_train_station(trainstops[station][0])
                    station_cache[trainstops[station][0]] = trainstops[station][1]
```

Reply:
We can use `station_cache[train[CONF_FROM]]` and station_cache[train[CONF_TO]]` instead of storing in variables.

Reply:
Haha yes! 🎉 

```
            trainstops = {"from_station": train[CONF_FROM],
                          "to_station": train[CONF_TO]}

            for station in trainstops:
                if trainstops[station] not in station_cache:
                    station_cache[trainstops[station]] = await train_api.async_get_train_station(trainstops[station])

....

        sensor = TrainSensor(train_api,
                             train[CONF_NAME],
                             station_cache[train[CONF_FROM]],
                             station_cache[train[CONF_TO]],
                             train[CONF_WEEKDAY],
                             train.get(CONF_TIME))
        sensors.append(sensor)

```

Reply:
Pushed, but changed the dict `trainstops` to an array.